azurelinuxagent/common/protocol/wire.py

```diff
@@ -806,42 +813,35 @@ def update_goal_state(self, force_update=False):
                 self._goal_state = goal_state
                 goal_state_updated = True
 
-            vm_settings_goal_state_updated = False
-            if not conf.get_enable_fast_track():
-                # if Fast Track is not enabled use extensionsConfig
-                self._extensions_goal_state = self._goal_state.extensions_config
+            #
+            # Now fetch the vmSettings (which contain the extensions goal state) from the HostGAPlugin
+            #
+            vm_settings_goal_state, vm_settings_goal_state_updated = (None, False)
+            vm_settings_goal_state_is_valid = False
+
+            if conf.get_enable_fast_track():
+                try:
+                    vm_settings_goal_state, vm_settings_goal_state_updated = self._fetch_vm_settings_goal_state(force_update=force_update)
+
+                    if goal_state_updated or vm_settings_goal_state_updated:
+                        # compare() raises a GoalStateMismatchError if the goal states don't match
+                        ExtensionsGoalState.compare(self._goal_state.extensions_config, vm_settings_goal_state)
+
+                    vm_settings_goal_state_is_valid = True
+
+                except Exception as error:
+                    # _fetch_vm_settings_goal_state() does its own detailed error reporting and raises ProtocolError; do not report those
+                    if not isinstance(error, ProtocolError):
+                        self._vm_settings_error_reporter.report_error(format_exception(error))
+                self._vm_settings_error_reporter.report_summary()
+
+            #
+            # If we fetched a valid vmSettings, use that for the extensions goal state. Otherwise use ExtensionsConfig
+            #
+            if vm_settings_goal_state_is_valid:
+                self._extensions_goal_state = vm_settings_goal_state
             else:
-                if not self._host_plugin_supports_vm_settings and self._host_plugin_supports_vm_settings_next_check > datetime.now():
-                    # if vmSettings are not supported use extensionsConfig
-                    self._extensions_goal_state = self._goal_state.extensions_config
-                else:
-                    try:
-                        vm_settings_goal_state = self._fetch_vm_settings_goal_state(force_update=force_update)
-
-                        self._host_plugin_supports_vm_settings = True
-
-                        if self._vm_settings_goal_state is None or self._vm_settings_goal_state.etag != vm_settings_goal_state.etag:
-                            self._vm_settings_goal_state = vm_settings_goal_state
-                            self._extensions_goal_state = vm_settings_goal_state
```

Comment:
Now self._extensions_goal_state is updated only once, after all the logic to decide which goal state to use has been executed